File: backend/app/models/wallet.py
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, field_validator

MIN_TOPUP = Decimal("1.00")
MAX_TOPUP = Decimal("1000.00")
DEFAULT_CURRENCY = "EUR"


class TopUpRequest(BaseModel):
    amount: Decimal

    @field_validator("amount")
    @classmethod
    def validate_amount(cls, v: Decimal) -> Decimal:
        if v < MIN_TOPUP:
            raise ValueError(f"Minimum top-up is {MIN_TOPUP}")
        if v > MAX_TOPUP:
            raise ValueError(f"Maximum top-up is {MAX_TOPUP}")
        return v.quantize(Decimal("0.01"))


class WalletResponse(BaseModel):
    balance: str
    currency: str
    updated_at: Optional[str] = None


class TransactionResponse(BaseModel):
    id: str
    type: str
    amount: str
    currency: str
    status: str
    provider: str
    provider_payment_id: Optional[str] = None
    created_at: str
    metadata: dict = {}


class TransactionListResponse(BaseModel):
    transactions: list[TransactionResponse]
    total: int
    page: int
    limit: int


class TopUpResponse(BaseModel):
    mode: str  # "stripe" | "mock"
    transaction_id: Optional[str] = None
    # Stripe-specific
    client_secret: Optional[str] = None
    payment_intent_id: Optional[str] = None
    publishable_key: Optional[str] = None
    # Mock — already completed
    status: Optional[str] = None
    balance: Optional[str] = None

File: frontend/src/api/wallet.ts
import Cookies from "js-cookie";

const apiUrl = process.env.REACT_APP_API_URL;

export interface WalletData {
  balance: string;
  currency: string;
  updated_at: string | null;
}

export interface TransactionData {
  id: string;
  type: string;
  amount: string;
  currency: string;
  status: string;
  provider: string;
  provider_payment_id: string | null;
  created_at: string;
  metadata: Record<string, unknown>;
}

export interface TransactionList {
  transactions: TransactionData[];
  total: number;
  page: number;
  limit: number;
}

export interface TopUpResult {
  mode: "stripe" | "mock";
  transaction_id?: string;
  // Stripe-specific
  client_secret?: string;
  payment_intent_id?: string;
  publishable_key?: string;
  // Mock — already completed
  status?: string;
  balance?: string;
}

function authHeaders(): Record<string, string> {
  const token = Cookies.get("access_token");
  return token
    ? { Authorization: `Bearer ${token}`, "Content-Type": "application/json" }
    : { "Content-Type": "application/json" };
}

export async function fetchWallet(): Promise<WalletData> {
  const res = await fetch(`${apiUrl}/wallet`, { headers: authHeaders() });
  if (!res.ok) throw new Error(`Failed to fetch wallet: ${res.status}`);
  return res.json();
}

export async function fetchTransactions(
  page = 1,
  limit = 20
): Promise<TransactionList> {
  const res = await fetch(
    `${apiUrl}/wallet/transactions?page=${page}&limit=${limit}`,
    { headers: authHeaders() }
  );
  if (!res.ok) throw new Error(`Failed to fetch transactions: ${res.status}`);
  return res.json();
}

export async function createTopUp(amount: number): Promise<TopUpResult> {
  const res = await fetch(`${apiUrl}/wallet/topup`, {
    method: "POST",
    headers: authHeaders(),
    body: JSON.stringify({ amount }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error((err as any).detail ?? `Top-up failed: ${res.status}`);
  }
  return res.json();
}
